include/ringmesh/mesh/surface_mesh.h

```diff
@@ -0,0 +1,416 @@
+         * @param[in] polygon_local_edge index of the polygon and
+         * the local edge starting vertex index
+         */
+        vecn< DIMENSION > polygon_edge_barycenter(
+            const PolygonLocalEdge& polygon_local_edge ) const;
+
+        /*!
+         * @brief Gets the vertex index on the polygon edge
+         * @param[in] polygon_local_edge index of the polygon and
+         * the local index of the edge in the polygon
+         * @param[in] vertex_id index of the local vertex in the edge \param
+         * edge_id (0 or 1)
+         * @return the vertex index
+         */
+        index_t polygon_edge_vertex( const PolygonLocalEdge& polygon_local_edge,
+            index_t vertex_id ) const;
+
+        /*!
+         * Computes the Mesh polygon barycenter
+         * @param[in] polygon_id the polygon index
+         * @return the polygon center
+         */
+        vecn< DIMENSION > polygon_barycenter( index_t polygon_id ) const;
+
+        /*!
+         * Computes the Mesh polygon area
+         * @param[in] polygon_id the polygon index
+         * @return the polygon area
+         */
+        virtual double polygon_area( index_t polygon_id ) const = 0;
+
+        /*!
+         * @brief return the NNSearch at polygons
+         */
+        const NNSearch< DIMENSION >& polygon_nn_search() const
+        {
+            if( !nn_search_ )
+            {
+                std::vector< vecn< DIMENSION > > polygon_centers(
+                    nb_polygons() );
+                for( auto p : range( nb_polygons() ) )
+                {
+                    polygon_centers[p] = polygon_barycenter( p );
+                }
+                nn_search_.reset(
+                    new NNSearch< DIMENSION >( polygon_centers, true ) );
+            }
+            return *nn_search_.get();
+        }
+        /*!
+         * @brief Creates an AABB tree for a Mesh polygons
+         */
+        const SurfaceAABBTree< DIMENSION >& polygon_aabb() const
+        {
+            if( !polygon_aabb_ )
+            {
+                polygon_aabb_.reset(
+                    new SurfaceAABBTree< DIMENSION >( *this ) );
+            }
+            return *polygon_aabb_;
+        }
+
+        bool is_mesh_valid() const override;
+
+        std::tuple< index_t, std::vector< index_t > >
+            connected_components() const final;
+
+    protected:
+        SurfaceMeshBase() = default;
+
+    private:
+        mutable std::unique_ptr< NNSearch< DIMENSION > > nn_search_{};
+        mutable std::unique_ptr< SurfaceAABBTree< DIMENSION > > polygon_aabb_{};
+
+        GEO::index_t find_first_polygon_around_vertex( GEO::index_t cur_p,
+            GEO::index_t vertex_id,
+            GEO::index_t& first_polygon ) const;
+        std::vector< index_t > store_polygons_around_vertex(
+            GEO::index_t first_polygon,
+            GEO::index_t vertex_id,
+            bool border_only ) const;
+        void stacking_polygons( GEO::index_t vertex_id,

```

Comment:
stacking polygons is not a good name

Reply:
The objective of this function is to add polygons into the vector result (not principally to the stack S)

Reply:
What about "push_polygons_from stack"?

Reply:
it is better

Reply:
Ok go for it